hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java

```diff
@@ -0,0 +1,160 @@
+    private final TaskDataSource source;
+
+    // mapper task controls
+    private final ForkJoinPool mapperPool;
+    private final int parallelism;
+    private final Semaphore enqueuePermits;
+
+    private long start = System.currentTimeMillis();
+    private long totalBundles = 0;
+
+    public MapFeederForkJoin(StreamMapper task, TaskDataSource source, int parallelism) {
+        this.source = source;
+        this.task = task;
+        this.parallelism = parallelism;
+        this.mapperPool = new ForkJoinPool(parallelism);
+        this.enqueuePermits = new Semaphore(QUEUE_DEPTH);
+    }
+
+    @Override public void run() {
+        log.info("starting {} thread(s) for src={}", parallelism, source);
+        try {
+            if (source.isEnabled()) {
+                while (fillBuffer()) {
+                    if (Thread.interrupted()) {
+                        closeSourceIfNeeded();
+                    }
+                }
+            }
+            closeSourceIfNeeded();
+            joinProcessors();
+            log.info("all ({}) task threads exited; sending taskComplete", parallelism);
+            logBundleThroughput();
+            // run in different threads to isolate them from interrupts. ie. "taskCompleteUninterruptibly"
+            // join awaits completion, is uninterruptible, and will propagate any exception
+            CompletableFuture.runAsync(task::taskComplete).join();
+        } catch (Throwable t) {
+            logBundleThroughput();
+            handleUncaughtThrowable(t);
+        }
+        log.debug("task feeder exited");
+    }
+
+    private void logBundleThroughput() {
+        long elapse = (System.currentTimeMillis() - start) / 1000;
+        log.info("{} bundles processed in {} seconds (avg rate={}/s)", totalBundles, elapse, totalBundles / elapse);
+    }
+
+    /**
+     * Immediately halting is the best way we have of ensuring errors are reported without a significantly
+     * more involved system.
+     */
+    private void handleUncaughtThrowable(Throwable t) {
+        if (errored.compareAndSet(false, true)) {
+            log.error("unrecoverable error in task feeder or one of its mapper threads. immediately halting jvm", t);
+            Runtime.getRuntime().halt(1);
+        }
+    }
+
+    /**
+     * Push the next bundle onto the work queue. Returns true
+     * to continue processing. If another thread has interrupted
+     * ourselves then set our interrupt status. This will close
+     * the current source and continue to consume any remaining elements
+     * from the queue.
+     */
+    private boolean fillBuffer() {
+        boolean status = false;
+        try {
+            enqueuePermits.acquire();
+            Bundle p = source.next();
+            if (p == null) {
+                log.info("exiting on null bundle from {}", source);
+            } else {
+                totalBundles++;
+                mapperPool.submit(new MapperTask(p));
```

Comment:
do we need to both create a task and then have the task do its own post-complete accounting? with the uncaught exception handler and a listener on the future, it can just be like a method reference